docs: gitlab-runner with example & VM test (shell-runner & podman-runner)

gabyx opened the pull request:
This documents the `gitlab-runner` service with a concret example which works out of the box.
The `gitlab-runner.nix` has been moved to `gitlab-runner/runner.nix` with a VM tests with the following runner

- `shell-runner`
- `podman-runner,` which is documented as an example in the NixOS manual under `gitlab` which comprises of a full blown `/nix/store` shared Gitlab runner which can be taken as reference and reused.

:heart:  NixCon25 was awesome and talking to @Nebucatnetzer gave me enthusiasm to share this documentation and extend it with a example which works. As CI with Nix is a really important cornerstone to make adoption greater. This is used here: https://gitlab.com/data-custodian/custodian
and took quite some while to figure out how to do it etc. We are using that setup since 1 year.
The NixOS wiki on the gitlab-runner is also outdated and not so nice.

**Status**:

- [x] VM test partially done. @Nebucatnetzer thanks for your kind help.
  - [x] Shell executor runner works.										
  - [x] Container executor.
- [x] Documentation partially done.
  - [x] Clean up etc and make CI happy.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [x] Module update: when the change is significant. (**no module changes but mention documentation for gitlab-runner etc**).
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

gabyx committed:
docs: gitlab-runner with example

xanderio commented:
Thanks for the contribution @gabyx, this is really awesome.

I'm not happy putting this in an example folder next to the NixOS module, you could kill two birds with one stone by turning this into a NixOS test. That way we can ensure that it still works and also have it at a place we people we look for it. As this would also require setting up a working GitLab instance inside the NixOS test, this would make the whole thing quite a bit more complex.

Maybe could just a [wiki](https://wiki.nixos.org) page, that way it will be discoverable and would require writing such a complex NixOS test.

Where do you want to take this? :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-run-rootless-containers-under-a-user-podman-in-nixos/69834/6

Nebucatnetzer commented on nixos/modules/services/continuous-integration/gitlab-runner/runner.md:
```diff
@@ -0,0 +1,72 @@
+
+  **For Gitlab Jobs**:
+
+  - an image based on Ubuntu (name: `local/alpine`) with a Nix installation used
```
There is a mixup with the names here.
You call the Ubuntu image `local/alpine` and vise versa.

Nebucatnetzer reviewed (commented):
The NixOS tests for GitLab live here: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/gitlab.nix

In general it would more sense to me to go with a more classic setup as the main example for setting up a GitLab Runner.

gabyx committed:
fix: add VM test for `serivces.gitlab-runner`

- Put all tests into `gitlab` folder.
- Add test for `gitlab-runner` service:
  - Add runner `shell-executor`.

gabyx commented:
@Nebucatnetzer: thanks introducing me to nixos VM tests, we have now a first draft of a working gitlab-runner vmTest.
I will in the next continuation add some different runners to the VM test from

-  simple one (just shell executor) for testing that gitlab-runner registers and works.
-  the more complex example with all bells and whistles, nix-daemon in *container* and mounted into job containers.

gabyx committed:
chore: correct some documentation and add runner refactoring

gabyx committed:
fix: refactor for more runner tests

gabyx committed:
fix: refactor into python script

gabyx committed:
fix: typo

gabyx committed:
fix: print statment

gabyx committed:
fix: add podman executor (not tested)

gabyx committed:
fix: tests with podman runner

gabyx committed:
fix: down scale the VM to 5Gbi (works)

gabyx committed:
fix: wait for nix/podman daemon containers

gabyx committed:
fix: update documentation

gabyx committed:
fix: disk size outside

gabyx commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 441161`
Commit: `38434ea85b9572fc5dee18baa7527c3a577dcc33`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

gabyx committed:
fix: documentation

gabyx committed:
fix: 10gb disk

gabyx committed:
fix: only run podman-runner test on x86_64-linux

SuperSandro2000 commented on nixos/tests/gitlab/runner/podman-runner/root/etc/containers/registries.conf.d/000-shortnames.conf:
To run the tests we surely do not need a full alias list.

SuperSandro2000 reviewed (commented):
I don't understand why we vendor so many configs. Can't we copy them from a package or their source or a package's source?

gabyx commented:
Ah jeah: We can! I will try to replace them! I first wanted to make it work.

gabyx committed:
fix: remove obsolete comments

gabyx committed:
fix: add comment how to obtain these files

gabyx committed:
fix: improve image sizes

- do not copy Nix store paths

gabyx commented:
@ofborg build nixosTests.gitlab.runner

gabyx committed:
doc: add explanation how to clean disk space

gabyx commented:
@alejandrosame: Could somebody review this PR and approve it to be merged?
It was quite some work, and now we have some more tests for **untested** modules + documentation.

Thanks alot.

fricklerhandwerk reviewed (approved):
@gabyx asked me to take a look, and I did, superficially. I agree with @xanderio, it's an awesome addition and it seems like all comments were addressed. So, approved (modulo the merge conflict).

gabyx committed:
sync: with master

gabyx committed:
docs: typos corrections

gabyx committed:
fix: remove redundant file

gabyx committed:
fix: correct systemd service name for job images

gabyx committed:
fix: improvements on the images

gabyx committed:
fix: improve images

gabyx committed:
fix: add images to registry on startup

gabyx committed:
fix: volumes from flag

drupol reviewed (approved):
💯 Nice work !

Merging this tomorrow unless objections.

vcunat commented:
The CI only checks a very small fraction of things.

vcunat commented:
BTW, parts of the CI check names from commit messages before the colon, but given how you named them it checked: chore, chore.passthru.tests, doc, doc.passthru.tests, docs, docs.passthru.tests, fix, fix.passthru.tests, sync, sync.passthru.tests
